src/x/tls/config_manager.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tls
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/x/instrument"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
+
+const getConfigMetricName = "get-tls-config"
+
+var sleepFn = time.Sleep
+
+type ConfigManager interface {
+	TLSConfig() (*tls.Config, error)
+	ServerMode() ServerMode
+	ClientEnabled() bool
+}
+
+type configManager struct {
+	sync.RWMutex
+
+	log       *zap.Logger
+	metrics   configManagerMetrics
+	options   Options
+	certPool  *x509.CertPool
+	tlsConfig *tls.Config
+}
+
+type configManagerMetrics struct {
+	getTLSConfigSuccess tally.Counter
+	getTLSConfigErrors  tally.Counter
+}
+
+func newConfigManagerScope(scope tally.Scope) configManagerMetrics {
+	return configManagerMetrics{
+		getTLSConfigSuccess: scope.Tagged(map[string]string{"success": "true"}).Counter(getConfigMetricName),
+		getTLSConfigErrors:  scope.Tagged(map[string]string{"success": "false"}).Counter(getConfigMetricName),
+	}
+}
+
+func NewConfigManager(opts Options, instrumentOpts instrument.Options) ConfigManager {
+	scope := instrumentOpts.MetricsScope()
+	c := &configManager{
+		log:      instrumentOpts.Logger(),
+		metrics:  newConfigManagerScope(scope),
+		options:  opts,
+		certPool: x509.NewCertPool(),
+	}
+	go c.updateCertificates()
+	return c
+}
+
+func (c *configManager) updateCertificates() {
+	if c.options.CertificatesTTL() == 0 {
+		return
+	}
+	for {
```

Comment:
Seems like this should be a ticker. You can use the https://pkg.go.dev/github.com/benbjohnson/clock package, which has an interface for that.